silhouette/app/com/mohiva/play/silhouette/impl/providers/totp/GoogleTotpProvider.scala

```diff
@@ -0,0 +1,202 @@
+            logger.debug(VerificationCodeMustBeANumber.format(id))
+            false
+          case _ =>
+            logger.debug(VerificationCodeMustNotBeNullOrEmpty.format(id))
+            false
+        }
+      case _ =>
+        logger.debug(SharedKeyMustNotBeNullOrEmpty.format(id))
+        false
+    }
+  }
+
+  /**
+   * Returns the generated TOTP credentials including the shared key along with hashed scratch codes
+   * for safe storage, plain text scratch codes for first time use and the url to the QR activation code.
+   *
+   * @param accountName A unique key which identifies a user on this provider (userID, email, ...).
+   * @param issuer The issuer name. This parameter cannot contain the colon
+   * @return the generated TOTP credentials including the shared key, scratch codes and qr url.
+   */
+  def createCredentials(accountName: String, issuer: Option[String] = None): GoogleTotpCredentials = {
+    val credentials = googleAuthenticator.createCredentials()
+    val qrUrl = GoogleAuthenticatorQRGenerator.getOtpAuthURL(issuer.orNull, accountName, credentials)
+    val currentHasher = passwordHasherRegistry.current
+    val scratchCodesPlain = credentials.getScratchCodes.asScala.map(_.toString).toSeq
+    val hashedScratchCodes = scratchCodesPlain.map { scratchCode =>
+      currentHasher.hash(scratchCode)
+    }.toSeq
+    GoogleTotpCredentials(GoogleTotpInfo(credentials.getKey, hashedScratchCodes), scratchCodesPlain, qrUrl)
+  }
+
+  /**
+   * Returns some login info when the TOTP authentication with verification code was successful,
+   * None otherwise.
+   *
+   * @param sharedKey A unique key which identifies a user on this provider (userID, email, ...).
+   * @param verificationCode the verification code generated using TOTP.
+   * @return Some login info if the authentication was successful, None otherwise.
+   */
+  def authenticate(sharedKey: String, verificationCode: String): Future[Option[LoginInfo]] = {
+    Future(
+      if (isVerificationCodeValid(sharedKey, verificationCode)) {
+        Some(LoginInfo(ID, sharedKey))
+      } else {
+        logger.debug(VerificationCodeDoesNotMatch)
+        None
+      }
+    )
+  }
+
+  /**
+   * Authenticate the user using a TOTP scratch (or recovery) code. This method will
+   * check each of the previously hashed scratch codes and find the first one that
+   * matches the one entered by the user. The one found is removed from `totpInfo` and returned
+   * for easy client-side bookkeeping.
+   *
+   * @param totpInfo The original TOTP info containing the hashed scratch codes.
+   * @param plainScratchCode The plain scratch code entered by the user.
+   * @return Some tuple consisting of (`PasswordInfo`, `TotpInfo`) if the authentication was successful, None otherwise.
+   */
+  def authenticate(totpInfo: GoogleTotpInfo, plainScratchCode: String): Future[Option[(PasswordInfo, GoogleTotpInfo)]] = Future {
+    Option(totpInfo).flatMap { totpInfo =>
```

Comment:
Why is there an option? To handle a null value?

Reply:
@akkie I guess it's null check. I didn't change it yet, waiting for @bravegag answer. I guess we don't need that, at least you don't have this kind of checks in your code (e.g com.mohiva.play.silhouette.impl.providers.CredentialsProvider#authenticate)